File: levelorder/new_level_order.rb
class TreeNode

  attr_accessor :val, :left, :right

  def initialize(val)
    @val = val
    @left = @right = nil
  end
end

def level_order(root)
  return [] if root.nil?

  result = []
  queue = []
  queue << root
  while queue.length > 0
    current = []
    size = queue.size
    i = 0
    while i < size
      current_node = queue.shift
      current << current_node.val
      queue << current_node.left if current_node.left
      queue << current_node.right if current_node.right
      i += 1
    end
    result << current
  end
  result
end

root = TreeNode.new(5)
root.left = TreeNode.new(3)
root.right = TreeNode.new(7)
root.left.left = TreeNode.new(2)
root.left.right = TreeNode.new(4)
root.right.left = TreeNode.new(6)
root.right.right = TreeNode.new(8)
p level_order(root)

File: levelorder/newLevelOrder.js
//  this version separates out each level in separate sub arrays

  // Definition for a binary tree node.
  function TreeNode(val) {
      this.val = val;
      this.left = this.right = null;
  }


var levelOrder = function(root) {
  if (!root) {
    return;
  }
  let result = [];
  let stack = [];


  stack.push(root);
  while (stack.length > 0) {
    let current = [];
    let size = stack.length;
    for (let i = 0; i < size; i++) {
      let currentNode = stack.shift();
      current.push(currentNode.val);

      if (currentNode.left) {
        stack.push(currentNode.left);
      }

      if (currentNode.right) {
        stack.push(currentNode.right);
      }
    }
    result.push(current);
  }
  return result;
};

//      4
//   2    6
// 1  3  5  7
let node = new TreeNode(4);
node.left = new TreeNode(2);
node.right = new TreeNode(6);
node.left.left = new TreeNode(1);
node.left.right = new TreeNode(3);
node.right.left = new TreeNode(5);
node.right.right = new TreeNode(7);
console.log(levelOrder(node));
